ext/DeviceUtilsCUDAExt.jl

```diff
@@ -0,0 +1,122 @@
+module DeviceUtilsCUDAExt
+
+using Adapt: Adapt
+using CUDA: CUDA
+using CUDA.CUSPARSE: AbstractCuSparseMatrix, AbstractCuSparseVector
+using DeviceUtils: DeviceUtils, CUDADevice, CPUDevice, reset_gpu_device!
+using Random: Random
+
+
+__init__() = reset_gpu_device!()
+
+const USE_CUDA_GPU = Ref{Union{Nothing, Bool}}(nothing)
+
+function _check_use_cuda!()
+    USE_CUDA_GPU[] === nothing || return
+
+    USE_CUDA_GPU[] = CUDA.functional()
```

Comment:
CUDA devices should be selected only if cuDNN is functional. Else it creates all kinds of issues, re hitting scalar indexing in NNlib CPU versions.

(And users tend to never read warnings, so the cudnn not loaded will almost certainly be ignored)

Reply:
done, although it is a pity requiring cuDNN when it is not needed for data movement since there are non-deep-learning applications that could just do with CUDA.jl. Enforcing the presence of cuDNN should be done by DL frameworks.

Reply:
I see the confusion; see my comment above.